neo/rawio/axonarawio.py

```diff
@@ -0,0 +1,348 @@
+"""
+This class reads .set and .bin file data from the Axona acquisition system.
+
+File format overview:
+http://space-memory-navigation.org/DacqUSBFileFormats.pdf
+
+In brief:
+ data.set - setup file containing all hardware setups related to the trial
+ data.bin - raw data file
+
+There are many other data formats from Axona, which we do not consider (yet).
+These are derivative from the raw continuous data (.bin) and could in principle
+be extracted from it (see file format overview for details).
+
+Author: Steffen Buergers
+
+"""
+
+from .baserawio import (BaseRawIO, _signal_channel_dtype, _signal_stream_dtype,
+                        _spike_channel_dtype, _event_channel_dtype,
+                        _common_sig_characteristics)
+import numpy as np
+import os
+import re
+import contextlib
+import datetime
+
+
+class AxonaRawIO(BaseRawIO):
+    """
+    Class for reading raw, continuous data from the Axona dacqUSB system:
+    http://space-memory-navigation.org/DacqUSBFileFormats.pdf
+
+    The raw data is saved in .bin binary files with an accompanying
+    .set file about the recording setup (see the above manual for details).
+
+    Usage:
+        import neo.rawio
+        r = neo.rawio.AxonaRawIO(
+            filename=os.path.join(dir_name, base_filename)
+        )
+        r.parse_header()
+        print(r)
+        raw_chunk = r.get_analogsignal_chunk(block_index=0, seg_index=0,
+                      i_start=0, i_stop=1024,  channel_names=channel_names)
+        float_chunk = reader.rescale_signal_raw_to_float(
+            raw_chunk, dtype='float64',
+            channel_indexes=[0, 3, 6]
+        )
+    """
+
+    extensions = ['bin']  # Never used?
+    rawmode = 'one-file'
+
+    # In the .bin file, channels are arranged in a strange order.
+    # This list takes a channel index as input and returns the actual
+    # offset for the channel in the memory map (self._raw_signals).
+    channel_memory_offset = [
+        32, 33, 34, 35, 36, 37, 38, 39, 0, 1, 2, 3, 4, 5,
+        6, 7, 40, 41, 42, 43, 44, 45, 46, 47, 8, 9, 10, 11,
+        12, 13, 14, 15, 48, 49, 50, 51, 52, 53, 54, 55, 16, 17,
+        18, 19, 20, 21, 22, 23, 56, 57, 58, 59, 60, 61, 62, 63,
+        24, 25, 26, 27, 28, 29, 30, 31
+    ]
+
+    def __init__(self, filename):
+        BaseRawIO.__init__(self)
+
+        # We accept base filenames, .bin and .set extensions
+        self.filename = filename.replace('.bin', '').replace('.set', '')
+        self.bin_file = os.path.join(self.filename + '.bin')
+        self.set_file = os.path.join(self.filename + '.set')
+        self.set_file_encoding = 'cp1252'
+
+    def _source_name(self):
+        return self.filename
+
+    def _parse_header(self):
+        '''
+        Read important information from .set header file, create memory map
+        to raw data (.bin file) and prepare header dictionary in neo format.
+        '''
+
+        # Get useful parameters from .set file
+        params = ['rawRate']
```

Comment:
Just was checking w/ Abid on this project and it popped up in my head. There is this looming bug in the Axona software where it will save the sample rate at 48kHz even if the end-user recorded in their 2ms spike (24kHz) mode. You have to do an additional check for the `Spike2msMode` parameter. If the value is 0, then it is 48k, else 24k.

An additional caveat, I saw that Abid sent you a 2ms mode file that was just converted from the RAW mode. That converter I think will just dump the same `.set` file, so 2ms mode will not be set to 1 (thus the above check isn't even true in the case where the end-user wants to chunk their data from the raw mode to the 24k mode, although I'm not sure why they would want to). 

Could try to just infer sampling rate from the timestamps of the `.N` files. I think a majority of the end-users use the 48k mode though.

Reply:
Hi Geoff,

Thanks a bunch for pointing this out! We will take it into account!

Cheers,
Steffen.

Reply:
Hi @GeoffBarrett. Thanks for the feedback. I updated the current AxonaIO PR accordingly: https://github.com/NeuralEnsemble/python-neo/pull/986
Now the `Spike2msMode` is first taken into account to determine the sample rate and only if this fails the `.N` file header `sample_rate` is used. For now I am not trying to extract the sample rate from the timestamp resolution as I think this method also won't be 100% reliable, depending on the number of spikes extracted and the underlying precision of the timestamps.

Reply:
@JuliaSprenger you're right, the timestamps in the `.N` would just provide the time that the signal went above threshold, so they wouldn't really allow you to infer sampling frequency. I think the best you'd be able to do is read that `Spike2msMode`. I haven't used the pipeline, but potentially the end-user could provide a sample frequency override kwarg, not sure if it's some sort of CLI and the users can provide parameters or something. 

It is definitely an issue that I brought up with Jim Donnett a few years back. If you look through the header info in the 2ms data that Abid has provided you will see the 48k (as well as in the `.set`) so the issue is definitely there. I don't have any 24kHz recorded data, but you could always reach out to Abid and have him record a minute of 24kHz data (not converted from `.bin`), and look at the headers.